Continuing clean up/modernization of Performance docs (#2457)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/evaluate-performance/index.md

   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.  -->

> # Introduction to the Performance tool

Author: @mikehoffms, please do an edit pass on this doc and the second one that I'm working on. I'm flexible on the titles but before we had "Get started analyzing runtime performance" and "Analyze runtime performance" and I think we just need a clearer distinction. This one is really a tutorial/introduction/get started guide and I'd like the title to reflect that.

Reviewer: The std todo list for changing a title:
* YAML field at top of .md
* h1 near top of .md
* toc entry
* toc index entry possibly add the old title, comma delim'd (often overlooked, check this)
* full text search repo for old title string (likely overlooked; todo)

Reviewer: fixed 1x old "Get started analyzing..."

_Runtime performance_ is how your page performs when it's running, as opposed to loading.  The following tutorial teaches you how to use the DevTools **Performance** tool to analyze runtime performance.

The skills you learn in this tutorial are useful for analyzing loading, interactivity, and visual stability of your web content, which are also key indicators for [Core Web Vitals](https://web.dev/vitals/).  Each of the Core Web Vitals represents a distinct facet of the user experience, is measurable in the field, and reflects the real-world experience of a critical user-centric outcome.  You can see these Core Web Vitals in the Performance tool.

See also:

> - [Optimize website speed using Lighthouse](../speed/get-started.md).

Reviewer: ```suggestion
- [Optimize website speed using Lighthouse](../speed/get-started.md)

```
minor convention tweaks.  i put period if there's info to the right of link

Using the workflows and tools that you just learned, click **Optimize** on the demo to turn on the optimized code, take another performance recording, and then analyze the results.  From the improved framerate to the reduction in events in the flame chart in the **Main** section, the optimized version of the app does much less work, resulting in better performance.

> Compare this snippet of JavaScript from the unoptimized version of the app:

Author: @captainbrosset, please review this section for technical accuracy. I did my best to understand the differences between the optimized and unoptimized versions but feel free to make any corrections/suggestions as appropriate. Here is a link to the source code: https://github.com/MicrosoftEdge/Demos/blob/main/devtools-performance-get-started/app.js

Reviewer 2: Sure, I'll leave comments below if I find anything.

I think we should also update the demo app code. I find it hard to read personally. We could make it a lot more readable and decorated with comments.
If I did this, would you be ok updating these docs again? I'm wondering if we should just leave it aside for now until you're done with your clean-up PRs, or do all at the same time.

Reviewer: A few months ago, I inspected the code, and was not able to understand the optimization.

Author: I'm fine with parallelizing the demo app code changes and this PR. We can always land a future PR for this doc once the demo app code is updated.

Reviewer: @captainbrosset - maybe enter work item:
update the demo app code. make it a lot more readable and decorated with comments.

```javascript
var pos = m.classList.contains("down")
  ? m.offsetTop + distance
  : m.offsetTop - distance;
if (pos < 0) pos = 0;
if (pos > maxHeight) pos = maxHeight;
m.style.top = pos + "px";
if (m.offsetTop === 0) {
  m.classList.remove("up");
  m.classList.add("down");
}
if (m.offsetTop === maxHeight) {
  m.classList.remove("down");
  m.classList.add("up");
}
```

> `m` references each blue icon on the page. In the unoptimized version, we refer to `pos` as the new vertical position of the icon, calculated based on `m.offsetTop + distance` if the icon is moving down and where `distance` is a fixed increment.  If the icon is moving up instead, the new position is instead calculated with `m.offsetTop - distance`.
> 
> We check if we've exceeded either `0` or `maxHeight` as the new position whether we're moving up or down, adjust accordingly, and then we finally set the new position for each icon in CSS with `m.style.top = pos + "px";`.

Reviewer 2: ```suggestion
The above code snippet runs on every frame of the browser rendering loop and for each blue icon on the page. The `m` variable references a single blue icon.

In this unoptimized version, we create a `pos` variable that's set to the current position of the icon to which we add some distance. The current position of the icon is read by using `m.offsetTop`.

After making sure that the icon is still within the bounds of the page, we set its new position by using `m.style.top` which sets inline styles on the element.

Finally, we read `m.offsetTop` again to adjust the direction of the icon.
```

I'm hoping this helps clarify the code snippet a bit more. I removed some of the up/down details since the really important parts here are how we read and write the position of the icon.

Reviewer: I'm wondering if this is where I asked to pls add a direct statement that positively states what the optimization is/does.  I think my ask was re: this article.  Does this PR w/ this change, now have a simple, direct, positive summary of what the optimization is?

Reviewer: i would add comma:
```
that's set to the current position of the icon, to which we add some distance.
```

Reviewer: ``
we set its new position by using `m.style.top`, which sets inline styles on the element.
```
I would add comma, per style guide.
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/that-vs-which

Reviewer: add comma:
```
read `m.offsetTop` again, to adjust the direction of the icon.
```

> The optimized code uses a different sequence of actions to do less work. Here is the same snippet of JavaScript from the optimized version of the app: 

Reviewer: global, as many as 7 spots: change "unoptimized" to "non-optimized"?  I lean towards non- but not sure how to justify via style guides.  Does UI say "unoptimized" on a button label?

Author: <img width="235" alt="image" src="https://user-images.githubusercontent.com/34316555/222297087-aafc243e-0c85-44d1-8556-95d32e17e569.png">

When you click the "Optimize" button in the demo site, the button changes to say "Un-optimize". I can match the button or we can change what the button says since @captainbrosset was proposing updating the demo code to make it more readable.

Reviewer: There seems to be way more web search hits on 'unoptimized' than "non-optimized".  I'd encourage going with whatever phrasing actual users in the field/industry would actually write & say & recognize.

```javascript
var pos = parseInt(m.style.top.slice(0, m.style.top.indexOf("px")));
m.classList.contains("down") ? (pos += distance) : (pos -= distance);
if (pos < 0) pos = 0;
if (pos > maxHeight) pos = maxHeight;
m.style.top = pos + "px";
if (pos === 0) {
  m.classList.remove("up");
  m.classList.add("down");
}
if (pos === maxHeight) {
  m.classList.remove("down");
  m.classList.add("up");
}
```

> In the optimized version, `pos` is used instead of `m.offsetTop`. The remaining work is the same but this snippet is faster than the previous one because we are no longer querying the style of each icon while also changing it, which triggers the browser to re-layout each icon.  

Reviewer 2: ```suggestion
In the optimized version, we first set the value of the `pos` variable by reading `m.style.top` instead of using `m.offsetTop`. Using the element's inline style is faster because reading `m.offsetTop` forces the browser engine to know where all the elements are on the page, which requires the engine to recalculate the styles and the layout.

We then set the new position of the icon like the previous version, but we don't read `m.offsetTop` again like we did before to adjust the icon's direction.

The unoptimized code reads and writes the position of the icon from two different places, forcing the browser to recalculate the style and layout on each frame. The optimized version, however, writes and reads the position of the icon in the inline styles only.
```

I'm not completely sure if this is clear to readers, let me know what you think.
In any case, your  understanding was correct. So feel free to use my suggestions or not.

Again, I think it'd be great to recreate this demo, and make the optimized version even more optimized so as to make more obviously different from the unoptimized version. Right now they're quite confusingly similar.

Reviewer: At a glance, it looks like this addresses my previous ask (a few months ago), for a direct positive description of what the optzn is.

> Note that this code could be made even faster by only using properties that only affect compositing, instead of manipulating the `top` property of every icon.  For example, we could get even better performance if instead of using `m.style.top = pos + "px";`, we used `m.style.transform = translateY(pos + "px,");`.  This is because `transform: translate(npx, npx)` is a CSS attribute that only affects the compositor layer and doesn't trigger re-layout for the browser.  For more information, see [Use transform and opacity changes for animations](https://web.dev/stick-to-compositor-only-properties-and-manage-layer-count/#use-transform-and-opacity-changes-for-animations).

Reviewer 2: How about actually implementing this in the optimized version?

```suggestion
Note that this code could be made even faster by using CSS properties that only require the browser to do compositing rather than layout. Instead of manipulating the `top` property which forces the browser to run its layout code again, using the `transform` property would allow the browser to consider each icon as individual layers and display them in the right places by re-compositing the final image. For example, instead of using `m.style.top = pos + "px";`, we can use `m.style.transform = translateY(pos + "px,");`.

To learn more, see [Use transform and opacity changes for animations](https://web.dev/stick-to-compositor-only-properties-and-manage-layer-count/#use-transform-and-opacity-changes-for-animations).
```

Reviewer: more gramm'l: consistency of singular/plural (I favor singular, for least ambiguity of many-to-1 situations):

to consider **each** icon as **an** individual layer       

pls tech review that wording

Reviewer: imo, need a comma near the end:

```
Note that this code could be made even faster by using CSS properties 
that only require the browser to do compositing, rather than layout.
```

  m.classList.add("up");
}
```

The above code snippet runs on every frame of the browser rendering loop, for each blue icon on the page.  The `m` variable references a single blue icon.

In this unoptimized version, we create a `pos` variable that's set to the current position of the icon, to which we add some distance.  The current position of the icon is read by using `m.offsetTop`.

After making sure that the icon is still within the bounds of the page, we set its new position by using `m.style.top`, which sets inline styles on the element.

Finally, we read `m.offsetTop` again, to adjust the direction of the icon.

The optimized code uses a different sequence of actions to do less work. Here is the same snippet of JavaScript from the optimized version of the app: 

```javascript
var pos = parseInt(m.style.top.slice(0, m.style.top.indexOf("px")));
m.classList.contains("down") ? (pos += distance) : (pos -= distance);
if (pos < 0) pos = 0;
if (pos > maxHeight) pos = maxHeight;
m.style.top = pos + "px";
if (pos === 0) {
  m.classList.remove("up");
  m.classList.add("down");
}
if (pos === maxHeight) {
  m.classList.remove("down");
  m.classList.add("up");
}
```

> In the optimized version, we first set the value of the `pos` variable by reading `m.style.top` instead of using `m.offsetTop`.  Using the element's inline style is faster because reading `m.offsetTop` forces the browser engine to know where all the elements are on the page, which requires the engine to recalculate the styles and the layout.

Reviewer: ```suggestion
In the optimized version, we first set the value of the `pos` variable by reading `m.style.top` instead of using `m.offsetTop`.  Using the element's inline style is faster, because reading `m.offsetTop` forces the browser engine to know where all the elements are on the page, which requires the engine to recalculate the styles and the layout.
```
i am of the persuasion of using commas whenever possible, to assist parsing and guide where one would pause when speaking aloud to establish separate phrase-boundaries.  discussed previously with r.a. and we agreed on this philosophy.  up to you.

Reviewer: added comma

> We then set the new position of the icon like the previous version, but we don't read `m.offsetTop` again like we did before to adjust the icon's direction.

Reviewer: ```suggestion
We then set the new position of the icon, like done in the unoptimized version, but we don't read `m.offsetTop` again, like done in the unoptimized version, to adjust the icon's direction.
```
tech review this change, re: which scenario we're describing

Reviewer: incorp'd and added headings:
```
#### Unptimized version
#### Optimized version
```

The unoptimized code reads and writes the position of the icon from two different places, forcing the browser to recalculate the style and layout on each frame. The optimized version, however, writes and reads the position of the icon in the inline styles only.

> Note that this code could be made even faster by using CSS properties that only require the browser to do compositing, rather than layout.  Instead of manipulating the `top` property which forces the browser to run its layout code again, using the `transform` property would allow the browser to consider each icon as individual layers and display these layers in the right positions by re-compositing the final image.  For example, instead of using `m.style.top = pos + "px";`, we can use `m.style.transform = translateY(pos + "px,");`.

Reviewer: ```suggestion
Note that this code could be made even faster by using CSS properties that only require the browser to do compositing, rather than layout.  Instead of manipulating the `top` property, which forces the browser to run its layout code again, using the `transform` property would allow the browser to consider each icon as an individual layer, and then display these layers in the correct positions by re-compositing the final image.  For example, instead of using `m.style.top = pos + "px";`, we can use `m.style.transform = translateY(pos + "px,");`.
```
several wording changes for your consideration, as you see fit

Reviewer: incorp'd
